wrappers/kmip/kmip.go

```diff
@@ -0,0 +1,605 @@
+	if k.serverName != "" {
+		wrapConfig.Metadata["server_name"] = k.serverName
+	}
+
+	return wrapConfig, nil
+}
+
+// Type returns the type for this particular wrapper implementation
+func (k *Wrapper) Type(_ context.Context) (wrapping.WrapperType, error) {
+	return wrapping.WrapperTypeKmip, nil
+}
+
+// KeyId returns the last known key id
+func (k *Wrapper) KeyId(_ context.Context) (string, error) {
+	return k.currentKeyId.Load().(string), nil
+}
+
+// Encrypt is used to encrypt the master key using the the KMIP.
+// This returns the ciphertext, and/or any errors from this
+// call. This should be called after the KMS client has been instantiated.
+func (k *Wrapper) Encrypt(_ context.Context, plaintext []byte, opt ...wrapping.Option) (*wrapping.BlobInfo, error) {
+	if plaintext == nil {
+		return nil, fmt.Errorf("given plaintext for encryption is nil")
+	}
+
+	env, err := wrapping.EnvelopeEncrypt(plaintext, opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error wrapping data: %w", err)
+	}
+
+	WrappedKey, err := k.client.EncryptDEK(context.Background(), nil, env.Key)
+	if err != nil {
+		return nil, fmt.Errorf("error encrypting data: %w", err)
+	}
+
+	// Store the current key id.
+	k.currentKeyId.Store(k.keyId)
+
+	ret := &wrapping.BlobInfo{
+		Ciphertext: env.Ciphertext,
+		Iv:         env.Iv,
+		KeyInfo: &wrapping.KeyInfo{
+			KeyId:      k.keyId,
+			WrappedKey: WrappedKey,
+		},
+	}
+
+	return ret, nil
+}
+
+// Decrypt is used to decrypt the ciphertext. This should be called after Init.
+func (k *Wrapper) Decrypt(_ context.Context, in *wrapping.BlobInfo, opt ...wrapping.Option) ([]byte, error) {
+	if in == nil {
+		return nil, fmt.Errorf("given input for decryption is nil")
+	}
+
+	keyBytes, err := k.client.DecryptDEK(context.Background(), nil, in.KeyInfo.WrappedKey)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting data encryption key: %w", err)
+	}
+
+	envInfo := &wrapping.EnvelopeInfo{
+		Key:        keyBytes,
+		Iv:         in.Iv,
+		Ciphertext: in.Ciphertext,
+	}
+	plaintext, err := wrapping.EnvelopeDecrypt(envInfo, opt...)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting data: %w", err)
+	}
+
+	return plaintext, nil
+}
+
+type kmipKMS struct {
+	// standard KMIP configuration options
+	endpoint         string
+	tlsConfig        *tls.Config
+	uniqueIdentifier string
+	readTimeout      uint8
+	writeTimeout     uint8
+}
+
+// EncryptDEK uses the KMIP encrypt operation to encrypt the DEK.
+func (kms *kmipKMS) EncryptDEK(ctx context.Context, nonce []byte, plainDEK []byte) ([]byte, error) {
+	conn, err := kms.connect()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	var iv []byte
+	iv, err = uuid.GenerateRandomBytes(nonceSize)
+	if err != nil {
+		return nil, err
+	}
+	respMsg, decoder, uniqueBatchItemID, err := kms.send(conn,
+		kmip14.OperationEncrypt,
+		EncryptRequestPayload{
+			UniqueIdentifier: kms.uniqueIdentifier,
+			Data:             append(iv, plainDEK...),
```

Comment:
I think you should avoid prepending the IV to the plaintext. If you do so, the IV will be XORed with itself on first iteration, meaning that the first ciphertext block will be equal to `AES_encrypt(0)`, which itself is XORed with the next block, and so on.

The first ciphertext block is now always the encryption of the all-zero block. Since the first ciphertext block never depends on the actual plaintext, it is always the same for a given key.

This annihilate the benefits of the IV which becomes predictable.

Instead you should prepend the IV to the ciphertext after encryption